cli/tools/jupyter.rs

```diff
@@ -0,0 +1,893 @@
+    let zmq_msg = msg.serialize(&self.hmac_key);
+    self.socket.send(zmq_msg).await?;
+    Ok(())
+  }
+}
+
+struct DealerComm {
+  name: String,
+  conn_str: String,
+  session_id: String,
+  hmac_key: hmac::Key,
+  socket: zeromq::DealerSocket,
+}
+
+impl DealerComm {
+  pub fn new(
+    name: &str,
+    conn_str: String,
+    session_id: String,
+    hmac_key: hmac::Key,
+  ) -> Self {
+    println!("dealer '{}' connection: {}", name, conn_str);
+    Self {
+      name: name.to_string(),
+      conn_str,
+      session_id,
+      hmac_key,
+      socket: zeromq::DealerSocket::new(),
+    }
+  }
+
+  pub async fn connect(&mut self) -> Result<(), AnyError> {
+    self.socket.bind(&self.conn_str).await?;
+
+    Ok(())
+  }
+
+  pub async fn recv(&mut self) -> Result<ZmqMessage, AnyError> {
+    let msg = self.socket.recv().await?;
+    Ok(msg)
+  }
+
+  pub async fn send(&mut self, msg: Message) -> Result<(), AnyError> {
+    let zmq_msg = msg.serialize(&self.hmac_key);
+    self.socket.send(zmq_msg).await?;
+    Ok(())
+  }
+}
+
+async fn create_zmq_reply(name: &str, conn_str: &str) -> Result<(), AnyError> {
+  println!("reply '{}' connection string: {}", name, conn_str);
+
+  let mut sock = zeromq::RepSocket::new(); // TODO(apowers313) exact same as dealer, refactor
+  sock.monitor();
+  sock.bind(conn_str).await?;
+
+  loop {
+    let msg = sock.recv().await?;
+    dbg!(&msg);
+    println!("{} got packet!", name);
+  }
+}
+
+fn create_conn_str(transport: &str, ip: &str, port: u32) -> String {
+  format!("{}://{}:{}", transport, ip, port)
+}
+
+fn parse_zmq_packet(data: &ZmqMessage) -> Result<(), AnyError> {
+  let _delim = data.get(0);
+  let _hmac = data.get(1);
+  let header = data.get(2).unwrap();
+  let _parent_header = data.get(3);
+  let _metadata = data.get(4);
+  let _content = data.get(5);
+
+  println!("header:");
+  dbg!(header);
+  let header_str = std::str::from_utf8(header).unwrap();
+  let header_value: MessageHeader = serde_json::from_str(header_str).unwrap();
+  println!("header_value");
+  dbg!(&header_value);
+  // validate_header(&header_value)?;
+
+  Ok(())
+}
+
+fn hmac_sign(
+  key: &hmac::Key,
+  header: &str,
+  parent_header: &str,
+  metadata: &str,
+  content: &str,
+) -> String {
+  let mut hmac_ctx = hmac::Context::with_key(key);
+  hmac_ctx.update(header.as_bytes());
+  hmac_ctx.update(parent_header.as_bytes());
+  hmac_ctx.update(metadata.as_bytes());
+  hmac_ctx.update(content.as_bytes());
+  let tag = hmac_ctx.sign();
+  let sig = HEXLOWER.encode(tag.as_ref());
+  sig
+}
+
+fn hmac_verify(
+  key: &hmac::Key,
+  expected_signature: &[u8],
+  header: &[u8],
+  parent_header: &[u8],
+  metadata: &[u8],
+  content: &[u8],
+) -> Result<(), ring::error::Unspecified> {
+  let mut msg = Vec::<u8>::new();
+  msg.extend(header);
+  msg.extend(parent_header);
+  msg.extend(metadata);
+  msg.extend(content);
+  let sig = HEXLOWER.decode(expected_signature).unwrap();
```

Comment:
```suggestion
  let sig = HEXLOWER.decode(expected_signature)?;
```
unwrap() seems a little unsafe.